raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-            handleBeginQuorumEpochResponse(response.sourceId());
-        } else if (responseData instanceof EndQuorumEpochResponseData) {
-            // Nothing to do for now
-        } else if (responseData instanceof FindQuorumResponseData) {
-            handleFindQuorumResponse((FindQuorumResponseData) responseData);
+        ApiKeys apiKey = ApiKeys.forId(response.data.apiKey());
+        final boolean handledSuccessfully;
+
+        switch (apiKey) {
+            case FETCH_QUORUM_RECORDS:
+                handledSuccessfully = handleFetchQuorumRecordsResponse(response);
+                break;
+
+            case VOTE:
+                handledSuccessfully = handleVoteResponse(response);
+                break;
+
+            case BEGIN_QUORUM_EPOCH:
+                handledSuccessfully = handleBeginQuorumEpochResponse(response);
+                break;
+
+            case END_QUORUM_EPOCH:
+                handledSuccessfully = handleEndQuorumEpochResponse(response);
+                break;
+
+            case FIND_QUORUM:
+                handledSuccessfully = handleFindQuorumResponse(response);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Received unexpected response type: " + apiKey);
+        }
+
+        ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (handledSuccessfully) {
+            connection.onResponseReceived(response.correlationId);
         } else {
-            throw new IllegalStateException("Received unexpected response " + response);
+            connection.onResponseError(response.correlationId, currentTimeMs);
         }
     }
 
-    private Optional<Exception> validateLeaderOnlyRequest(int remoteNodeId, int requestEpoch) throws IOException {
-        if (quorum.isObserver()) {
-            return Optional.of(Errors.INVALID_REQUEST.exception());
-        } else if (!quorum.isVoter(remoteNodeId)) {
-            return Optional.of(Errors.INVALID_REQUEST.exception());
-        } else if (requestEpoch < quorum.epoch()) {
-            return Optional.of(Errors.FENCED_LEADER_EPOCH.exception());
+    /**
+     * Validate a request which is only valid between voters.
+     */
+    private Optional<Errors> validateVoterOnlyRequest(int remoteNodeId, int requestEpoch) throws IOException {
+        if (requestEpoch < quorum.epoch()) {
+            return Optional.of(Errors.FENCED_LEADER_EPOCH);
+        } else if (quorum.isObserver() || !quorum.isVoter(remoteNodeId)) {
+            return Optional.of(Errors.INCONSISTENT_VOTER_SET);
         } else if (shutdown.get() != null) {
             shutdown.get().onEpochUpdate(requestEpoch);
-            return Optional.of(Errors.BROKER_NOT_AVAILABLE.exception());
+            return Optional.of(Errors.BROKER_NOT_AVAILABLE);
         }
 
-        // TODO: seems like a weird place to do this...
+        // The request is from a valid voter and indicates a larger epoch,
```

Comment:
This case combined with L881 looks weird, as if we don't know the identity of the sender, we shall early terminate; but in the meantime if it has a higher epoch, we shall just follow it. It might worth discussing what's the general strategy here, as even through static config, the quorum voters could change so the current receiver may hold stale membership data which prevents it from continuing to serve in the new quorum.

Reply:
I'll add a comment. As far as I understand it, the expected behavior in the reassignment protocol is to always trust the latest voter state we have seen in the log. But if we see a larger epoch from one of the expected voters, we need to take it.

Reply:
So the agreement is that sender will retry until it detects the latest membership info right?

Reply:
Yes, or until the receiver gets the latest membership if the sender is right.

Reply:
I was thinking about this a little more and it does seem we'll need to come back to it when we implement reassignment. The current validation makes sense only if you assume static voters. The problem is that a node which has been removed from the voter set needs to have some way to find the new voter set. It might not be able to do that if it only trusts its latest voter information. This might suggest the need for a new state, but not sure.